jep/0000/README.adoc

```diff
@@ -0,0 +1,345 @@
+
+| Created
+| Date (2019-05-15) :bulb:
+
+| BDFL-Delegate
+| TBD
+
+//
+//
+// Uncomment if there is an associated placeholder JIRA issue.
+//| JIRA
+//| :bulb: https://issues.jenkins-ci.org/browse/JENKINS-nnnnn[JENKINS-nnnnn] :bulb:
+//
+//
+// Uncomment if discussion will occur in forum other than jenkinsci-dev@ mailing list.
+//| Discussions-To
+//| :bulb: Link to where discussion and final status announcement will occur :bulb:
+//
+//
+// Uncomment if this JEP depends on one or more other JEPs.
+//| Requires
+//| :bulb: JEP-NUMBER, JEP-NUMBER... :bulb:
+//
+//
+// Uncomment and fill if this JEP is rendered obsolete by a later JEP
+//| Superseded-By
+//| :bulb: JEP-NUMBER :bulb:
+//
+//
+// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
+//| Resolution
+//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:
+
+|===
+
+== Abstract
+
+This JEP introduces a new system, which enables Jenkins plugin maintainers to add Continuous 
+Delivery to their projects. It is:
+
+* Opt-in: plugin maintainers can choose to use this
+* Simple to configure: very little needs to be changed in order to enable CD
+* Trusted CI/CD: deployments are produced from a single source of truth, which is managed by 
+the Jenkins project
+
+== Specification
+
+=== Overview
+
+We'll use a short story to illustrate what Autorelease will look like when a plugin maintainer 
+named Karl decides that he wants to perform continuous delivery of his (fictional) Karl-Plugin.
+
+To do so, Karl will:
+
+1. Verify support. Karl's plugin is already in the 
+link:https://github.com/jenkinsci[jenkinsci Organization on GitHub], and is 
+using `buildPlugin` when built on ci.jenkins.io.
+
+2. Choose a branch from which to release. Karl decides he wants to release from the `master` branch, 
+as that is the simplest.
+
+3. Settle on a version numbering scheme. Becuase it's in fairly wide use across the Jenkins 
+ecosystem, Karl settles on link:https://semver.org/[semantic versioning].
+
+4. Add a file to the top directory of his repository called `Autorelease.yaml`. 
+
+5. A webhook is configured on Karl's repository on GitHub to point to `{the-webhook-receiver}`
+
+Once ready to create a new release of his plugin, Karl merges some changes to the `master` branch. 
+At which point, the following takes place under the covers: 
+
+1. A webhook is sent to `{the-webhook-receiver}` by GitHub
+
+2. The receiver forwards this hook to `${trusted-ci-server}`, which is protected by a VPN
+
+3. `${trusted-ci-server}` uses the enhanced version of `buildPlugin` to validate the correctness of 
+Karl's `Autorelease.yaml` file, and the consistency of Karl's chosen version numbering.
+
+4. Once `Autorelease.yaml` has been validated, a plugin build is performed on `${trusted-ci-server}` 
+as one would expect.
+
+5. After the build is successful, `${trusted-ci-server}` handles deployment of the built plugin to 
+Artifactory and the Jenkins Update Center.
```

Comment:
To both? Or to Artifactory, which then causes the plugin to appear on the UC by virtue of the existing UC generator?

Reply:
Then I suppose to Artifactory. I didn't know that's how the UC gets populated. Will re-word to reflect that. Maybe 

> After the build is successful, the trusted CI server handles deployment of the built plugin to 
> Artifactory, at which point the existing UC generator makes the plugin available on the update center

If there's documentation on how that UC generator works, I'd (a) love to read that, and (b) be happy to have the words _existing UC generator_ link to that document.

Reply:
https://github.com/jenkins-infra/backend-update-center2 is the tool for it. We build and run it basically all the time. It downloads the periodically updated Maven index from Artifactory, and uses that to build up a list of plugins and their releases to deliver. There's a bunch of additional features like overrides and performance optimizations here, so it might look pretty weird. That it uses a really obsolete library to do the Maven index stuff doesn't help.

Basically, in terms of deployment, nothing needs to be different, server-side, from today when we get artifacts into Artifactory, the infra knows how to handle it. (Modulo the variable length version numbering scheme we talked about.)

Reply:
> nothing needs to be different

That we know of now.

Reply:
For what it's worth, I've added a link to the UC generator in the appropriate parts of Overview and Detailed Setup. If we find something that needs to be changed, we'll take that as it comes.